src/lib/optimizer/strategy/projection_combination_rule.cpp

```diff
@@ -0,0 +1,103 @@
+#include "projection_combination_rule.hpp"
+
+#include <memory>
+#include <string>
+#include <vector>
+
+#include "logical_query_plan/abstract_lqp_node.hpp"
+#include "logical_query_plan/projection_node.hpp"
+
+namespace opossum {
+
+std::string ProjectionCombinationRule::name() const { return "Projection Combination Rule"; }
+
+bool ProjectionCombinationRule::apply_to(const std::shared_ptr<AbstractLQPNode>& node) {
+  if (node->type() != LQPNodeType::Projection) {
+    return _apply_to_children(node);
+  }
+
+  // auto tree_altered = false;
+  std::vector<std::shared_ptr<ProjectionNode>> projection_nodes;
+
+  // Gather subsequent ProjectionNodes
+  auto current_node = node;
+  while (current_node->type() == LQPNodeType::Projection) {
+    projection_nodes.emplace_back(std::static_pointer_cast<ProjectionNode>(current_node));
+
+    // If current node has multiple or no children, end the combination here
+    if (current_node->right_child() || !current_node->left_child()) {
+      break;
+    }
+
+    current_node = current_node->left_child();
+
+    if (current_node->parents().size() > 1) {
+      break;
+    }
+  }
+
+  if (projection_nodes.size() > 1) {
+    const auto combined_projection_node = _combine_projections(projection_nodes);
+    _apply_to_children(combined_projection_node);
+    return true;
+  }
+
+  return _apply_to_children(node);
+}
+
+std::shared_ptr<ProjectionNode> ProjectionCombinationRule::_combine_projections(
+    std::vector<std::shared_ptr<ProjectionNode>>& projections) const {
+  // Store original child and parents
+  auto child = projections.back()->left_child();
+  const auto parents = projections.front()->parents();
+  const auto child_sides = projections.front()->get_child_sides();
+
+  auto column_expressions = std::vector<std::shared_ptr<LQPExpression>>();
+  for (const auto expression : projections.front()->column_expressions()) {
+    // If the expression is no reference to another column, we can just add it.
+    if (expression->type() != ExpressionType::Column) {
+      column_expressions.push_back(expression);
+      continue;
+    }
+
+    // If the expression IS a reference to another column,
+    // we have to check if the referenced column is part of one of the other ProjectionNodes that we are combining,
```

Comment:
Could you give an example of when this would be the case?

Reply:
```load src/test/tables/sqlite/int_int_int_100.tbl int_int_int; INSERT INTO int_int_int (id, int_a, int_b, int_c) SELECT 100, 1, 2, 3 FROM int_int_int;```

This creates two consecutive ProjectionNodes, one with the `100`, `1`, `2`, `3` literals, and one with LQPColumnReferences referencing those columns. So when we get rid of the child node, we also have to replace the references in the parent.

Reply:
So with "part of" you mean "created (e.g., literals, subselects) by a previous projection"?

Reply:
Yes.. Updating that comment